### ISSUE TITLE
CUDA compiler error

### ISSUE THREAD

**Issue by xx-zhou16** [traveller59/spconv#16]
error as following:
```
(python3.6) [zhouxiangxin@mad-wing ~/exp/3d_detection/experiments/spconv]$ sudo python3.6 setup.py bdist_wheel
running bdist_wheel
running build
running build_py
running build_ext
/home/zhouxiangxin/exp/3d_detection/experiments/spconv/build/lib.linux-x86_64-3.6
Release
-- The CUDA compiler identification is unknown
-- Check for working CUDA compiler: /unsullied/sharefs/_admin/cuda/cuda-9.0/cuda/bin/nvcc
-- Check for working CUDA compiler: /unsullied/sharefs/_admin/cuda/cuda-9.0/cuda/bin/nvcc -- broken
CMake Error at /usr/local/share/cmake-3.13/Modules/CMakeTestCUDACompiler.cmake:46 (message):
  The CUDA compiler

    "/unsullied/sharefs/_admin/cuda/cuda-9.0/cuda/bin/nvcc"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/zhouxiangxin/exp/3d_detection/experiments/spconv/build/temp.linux-x86_64-3.6/CMakeFiles/CMakeTmp

    Run Build Command:"/usr/bin/make" "cmTC_7a980/fast"
    /usr/bin/make -f CMakeFiles/cmTC_7a980.dir/build.make CMakeFiles/cmTC_7a980.dir/build
    make[1]: Entering directory '/home/zhouxiangxin/exp/3d_detection/experiments/spconv/build/temp.linux-x86_64-3.6/CMakeFiles/CMakeTmp'
    make[1]: Warning: File 'CMakeFiles/cmTC_7a980.dir/progress.make' has modification time 0.0067 s in the future
    Building CUDA object CMakeFiles/cmTC_7a980.dir/main.cu.o
    /unsullied/sharefs/_admin/cuda/cuda-9.0/cuda/bin/nvcc    "--expt-relaxed-constexpr"    -x cu -c /home/zhouxiangxin/exp/3d_detection/experiments/
spconv/build/temp.linux-x86_64-3.6/CMakeFiles/CMakeTmp/main.cu -o CMakeFiles/cmTC_7a980.dir/main.cu.o
    gcc: error trying to exec 'cc1plus': execvp: No such file or directory
    CMakeFiles/cmTC_7a980.dir/build.make:65: recipe for target 'CMakeFiles/cmTC_7a980.dir/main.cu.o' failed
    make[1]: *** [CMakeFiles/cmTC_7a980.dir/main.cu.o] Error 1
    make[1]: Leaving directory '/home/zhouxiangxin/exp/3d_detection/experiments/spconv/build/temp.linux-x86_64-3.6/CMakeFiles/CMakeTmp'
    Makefile:121: recipe for target 'cmTC_7a980/fast' failed
    make: *** [cmTC_7a980/fast] Error 2

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:2 (project)


-- Configuring incomplete, errors occurred!
See also "/home/zhouxiangxin/exp/3d_detection/experiments/spconv/build/temp.linux-x86_64-3.6/CMakeFiles/CMakeOutput.log".
See also "/home/zhouxiangxin/exp/3d_detection/experiments/spconv/build/temp.linux-x86_64-3.6/CMakeFiles/CMakeError.log".
Traceback (most recent call last):
  File "setup.py", line 86, in <module>
    zip_safe=False,
  File "/home/zhouxiangxin/anaconda3/envs/python3.6/lib/python3.6/site-packages/setuptools/__init__.py", line 143, in setup
    return distutils.core.setup(**attrs)
  File "/home/zhouxiangxin/anaconda3/envs/python3.6/lib/python3.6/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/home/zhouxiangxin/anaconda3/envs/python3.6/lib/python3.6/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/home/zhouxiangxin/anaconda3/envs/python3.6/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/home/zhouxiangxin/anaconda3/envs/python3.6/lib/python3.6/site-packages/wheel/bdist_wheel.py", line 188, in run
    self.run_command('build')
  File "/home/zhouxiangxin/anaconda3/envs/python3.6/lib/python3.6/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/home/zhouxiangxin/anaconda3/envs/python3.6/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/home/zhouxiangxin/anaconda3/envs/python3.6/lib/python3.6/distutils/command/build.py", line 135, in run
    self.run_command(cmd_name)
  File "/home/zhouxiangxin/anaconda3/envs/python3.6/lib/python3.6/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/home/zhouxiangxin/anaconda3/envs/python3.6/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "setup.py", line 39, in run
    self.build_extension(ext)
  File "setup.py", line 69, in build_extension
    subprocess.check_call(['cmake', ext.sourcedir] + cmake_args, cwd=self.build_temp, env=env)
  File "/home/zhouxiangxin/anaconda3/envs/python3.6/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)

subprocess.CalledProcessError: Command '['cmake', '/home/zhouxiangxin/exp/3d_detection/experiments/spconv', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/home/
zhouxiangxin/exp/3d_detection/experiments/spconv/build/lib.linux-x86_64-3.6/spconv', '-DCMAKE_PREFIX_PATH=/home/zhouxiangxin/anaconda3/envs/python3.
6/lib/python3.6/site-packages/torch', '-DPYBIND11_PYTHON_VERSION=3.6', '-DSPCONV_BuildTests=OFF', '-DCMAKE_CUDA_FLAGS="--expt-relaxed-constexpr"', '
-DCMAKE_BUILD_TYPE=Release']' returned non-zero exit status 1.

```
in the terminal, I can use `nvcc` normally.

**Comment by xx-zhou16** [comment]
which version of gcc/g++ is required?

**Comment by traveller59** [comment]
your gcc is imcomplete, the 'execvp' is part of g++. just google it and install.

**Comment by xx-zhou16** [comment]
> your gcc is imcomplete, the 'execvp' is part of g++. just google it and install.

Thank you for your reply.
Now the version of m gcc and g++ is 4.8.5.
I run `sudo python3.6 setup.py bdist_wheel`, and got errors as following:
```
/usr/include/c++/4.8/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C
++ 2011 standard. This support is currently experimental, and must be enabled with the -std=c++11 or -std=gnu++11 compiler
options.
 #error This file requires compiler and library support for the \
  ^
src/spconv/CMakeFiles/spconv.dir/build.make:88: recipe for target 'src/spconv/CMakeFiles/spconv.dir/indice.cu.o' failed
make[2]: *** [src/spconv/CMakeFiles/spconv.dir/indice.cu.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /home/zhouxiangxin/exp/3d_detection/experiments/spconv/include/spconv/spconv_ops.h:22:0,
                 from /home/zhouxiangxin/exp/3d_detection/experiments/spconv/src/spconv/indice.cc:17:
/home/zhouxiangxin/exp/3d_detection/experiments/spconv/include/torch_utils.h: In function ‘void tv::check_torch_dtype(const
 at::Tensor&)’:
/home/zhouxiangxin/exp/3d_detection/experiments/spconv/include/torch_utils.h:32:29: error: ‘remove_const_t’ is not a member
 of ‘std’
     auto val = std::is_same<std::remove_const_t<T>, double>::value;
```

**Comment by xx-zhou16** [comment]
And I tried build with `gcc/g++ 5.5.0`
And got errors as following:
```
[ 42%] Built target spconv_nms
[ 50%] Building CXX object src/spconv/CMakeFiles/spconv.dir/reordering.cc.o
/usr/lib/gcc/x86_64-linux-gnu/5/include/avx512fintrin.h(9220): error: argument of type "const void *" is incompatible with
parameter of type "const float *"

/usr/lib/gcc/x86_64-linux-gnu/5/include/avx512fintrin.h(9231): error: argument of type "const void *" is incompatible with
parameter of type "const float *"

/usr/lib/gcc/x86_64-linux-gnu/5/include/avx512fintrin.h(9244): error: argument of type "const void *" is incompatible with
parameter of type "const double *"

/usr/lib/gcc/x86_64-linux-gnu/5/include/avx512fintrin.h(9255): error: argument of type "const void *" is incompatible with
parameter of type "const double *"

/usr/lib/gcc/x86_64-linux-gnu/5/include/avx512fintrin.h(9268): error: argument of type "const void *" is incompatible with
parameter of type "const float *"

/usr/lib/gcc/x86_64-linux-gnu/5/include/avx512fintrin.h(9279): error: argument of type "const void *" is incompatible with
parameter of type "const float *"

/usr/lib/gcc/x86_64-linux-gnu/5/include/avx512fintrin.h(9292): error: argument of type "const void *" is incompatible with
parameter of type "const double *"

/usr/lib/gcc/x86_64-linux-gnu/5/include/avx512fintrin.h(9303): error: argument of type "const void *" is incompatible with
parameter of type "const double *"
```
which I think is caused by 5.5.0 gcc/g++

**Comment by bigsheep2018** [comment]
@xx-zhou16  check issue #5

**Comment by xx-zhou16** [comment]
I have checked my environment variable for `CUDA_TOOLKIT_ROOT`, but got this error as following:
```
Scanning dependencies of target spconv
Scanning dependencies of target spconv_nms
make[2]: Warning: File 'src/utils/CMakeFiles/spconv_nms.dir/depend.make' has modification time 0.029 s in the future
[  7%] Building CUDA object src/utils/CMakeFiles/spconv_nms.dir/nms.cu.o
make[2]: Warning: File 'src/spconv/CMakeFiles/spconv.dir/depend.make' has modification time 0.022 s in the future
[ 14%] Building CXX object src/spconv/CMakeFiles/spconv.dir/indice.cc.o
[ 28%] Building CUDA object src/spconv/CMakeFiles/spconv.dir/indice.cu.o
[ 28%] Building CXX object src/spconv/CMakeFiles/spconv.dir/all.cc.o
[ 35%] Linking CUDA device code CMakeFiles/spconv_nms.dir/cmake_device_link.o
[ 42%] Linking CUDA shared library ../../../lib.linux-x86_64-3.6/spconv/libspconv_nms.so
make[2]: warning:  Clock skew detected.  Your build may be incomplete.
[ 42%] Built target spconv_nms
Scanning dependencies of target spconv_utils
make[2]: Warning: File 'src/utils/CMakeFiles/spconv_utils.dir/depend.make' has modification time 0.031 s in the future
[ 50%] Building CXX object src/utils/CMakeFiles/spconv_utils.dir/all.cc.o
[ 57%] Building CXX object src/spconv/CMakeFiles/spconv.dir/reordering.cc.o
[ 64%] Building CUDA object src/spconv/CMakeFiles/spconv.dir/reordering.cu.o
[ 71%] Building CXX object src/spconv/CMakeFiles/spconv.dir/maxpool.cc.o
make[2]: *** No rule to make target '/usr/local/cuda/lib64/libculibos.a', needed by 'src/spconv/CMakeFiles/spconv.dir/cmake_device_link.o'.  Stop.
make[2]: *** Waiting for unfinished jobs....
[ 78%] Building CUDA object src/spconv/CMakeFiles/spconv.dir/maxpool.cu.o
[ 85%] Linking CXX shared library ../../../lib.linux-x86_64-3.6/spconv/spconv_utils.cpython-36m-x86_64-linux-gnu.so
make[2]: warning:  Clock skew detected.  Your build may be incomplete.
[ 85%] Built target spconv_utils
CMakeFiles/Makefile2:108: recipe for target 'src/spconv/CMakeFiles/spconv.dir/all' failed
make[1]: *** [src/spconv/CMakeFiles/spconv.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
Traceback (most recent call last):
  File "setup.py", line 86, in <module>
    zip_safe=False,
  File "/home/zhouxiangxin/anaconda3/lib/python3.6/site-packages/setuptools/__init__.py", line 129, in setup
    return distutils.core.setup(**attrs)
  File "/home/zhouxiangxin/anaconda3/lib/python3.6/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/home/zhouxiangxin/anaconda3/lib/python3.6/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/home/zhouxiangxin/anaconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/home/zhouxiangxin/anaconda3/lib/python3.6/site-packages/wheel/bdist_wheel.py", line 204, in run
    self.run_command('build')
  File "/home/zhouxiangxin/anaconda3/lib/python3.6/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/home/zhouxiangxin/anaconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/home/zhouxiangxin/anaconda3/lib/python3.6/distutils/command/build.py", line 135, in run
    self.run_command(cmd_name)
  File "/home/zhouxiangxin/anaconda3/lib/python3.6/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/home/zhouxiangxin/anaconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "setup.py", line 39, in run
    self.build_extension(ext)
  File "setup.py", line 70, in build_extension
    subprocess.check_call(['cmake', '--build', '.'] + build_args, cwd=self.build_temp)
  File "/home/zhouxiangxin/anaconda3/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--config', 'Release', '--', '-j4']' returned non-zero exit status 2.

```

**Comment by traveller59** [comment]
@xx-zhou16 this should be a cmake problem. the CMakeLists.txt use CMAKE_CUDA_COMPILER to determine the cuda location. you can print the CMAKE_CUDA_COMPILER.
you can delete the line 7~18 and remove ```${CUDA_CUDART} ${CUDA_CUBLAS}``` in line 30, but I can't test this.

**Comment by xx-zhou16** [comment]
Thank you so much. @traveller59  
Finally, I compile it successfully on CUDA 9.0, gcc/g++ 5.4.

**Comment by dhingratul** [comment]
@xx-zhou16 Can you summarize your findings, i have the same issue.
